File: ch02/stack_vm.py
"""
Stack VM - companion code for Chapter 2.

A fifteen-instruction stack machine.  Argument convention: caller pushes
arguments onto the operand stack before CALL; callee stores them into local
slots; return value is left on the operand stack at RET.

Entry point:
    StackVM(code).run()  ->  value at top of stack, or None
"""

# -- Opcodes

HALT  = 0   # stop; top of stack is the return value
PUSH  = 1   # push the next byte as an integer literal
ADD   = 2   # pop b, pop a; push a + b
SUB   = 3   # pop b, pop a; push a - b
MUL   = 4   # pop b, pop a; push a * b
DIV   = 5   # pop b, pop a; push a // b
EQ    = 6   # pop b, pop a; push 1 if a == b, else 0
LT    = 7   # pop b, pop a; push 1 if a < b, else 0
NOT   = 8   # pop a; push 0 if a is nonzero, else 1
JMP   = 9   # next two bytes: target address; jump
JMPZ  = 10  # next two bytes: target; jump if stack top zero
LOAD  = 11  # next byte: slot index; push locals[slot]
STORE = 12  # next byte: slot index; pop into locals[slot]
CALL  = 13  # next two bytes: target; save frame, jump
RET   = 14  # pop return value, restore frame, push it


# -- Core VM

class StackVM:
    def __init__(self, code):
        self.code   = code
        self.stack  = []        # operand stack
        self.frames = []        # saved frames: (locals, return_pc)
        self.locals = [0] * 16  # current frame's 16 local slots
        self.pc     = 0

    def _u16(self):             # read a big-endian 16-bit operand
        hi = self.code[self.pc]
        lo = self.code[self.pc + 1]
        self.pc += 2
        return (hi << 8) | lo

    def run(self):
        code = self.code
        while True:
            op = code[self.pc]; self.pc += 1

            if op == HALT:
                return self.stack[-1] if self.stack else None
            elif op == PUSH:
                self.stack.append(code[self.pc])
                self.pc += 1
            elif op == ADD:
                b = self.stack.pop(); self.stack[-1] += b
            elif op == SUB:
                b = self.stack.pop(); self.stack[-1] -= b
            elif op == MUL:
                b = self.stack.pop(); self.stack[-1] *= b
            elif op == DIV:
                b = self.stack.pop(); self.stack[-1] //= b
            elif op == EQ:
                b = self.stack.pop()
                self.stack[-1] = int(self.stack[-1] == b)
            elif op == LT:
                b = self.stack.pop()
                self.stack[-1] = int(self.stack[-1] < b)
            elif op == NOT:
                self.stack[-1] = int(not self.stack[-1])
            elif op == JMP:
                self.pc = self._u16()
            elif op == JMPZ:
                target = self._u16()
                if not self.stack.pop():
                    self.pc = target
            elif op == LOAD:
                s = code[self.pc]; self.pc += 1
                self.stack.append(self.locals[s])
            elif op == STORE:
                s = code[self.pc]; self.pc += 1
                self.locals[s] = self.stack.pop()
            elif op == CALL:
                target = self._u16()
                self.frames.append((self.locals[:], self.pc))
                self.locals = [0] * 16
                self.pc = target
            elif op == RET:
                val = self.stack.pop()
                self.locals, self.pc = self.frames.pop()
                self.stack.append(val)
            else:
                raise ValueError(f"unknown opcode {op} at pc={self.pc - 1}")


# -- Constant pool extension
# The book presents this as a sketch showing the one new branch to add.
# Here it is as a complete runnable class.

LOADK = 15  # next byte: index into constant pool

class StackVMWithPool(StackVM):
    """StackVM with a constant pool for values that do not fit in one byte."""

    def __init__(self, code, constants):
        super().__init__(code)
        self.constants = constants  # list of int, float, or str

    def run(self):
        code = self.code
        while True:
            op = code[self.pc]; self.pc += 1

            if op == LOADK:
                idx = code[self.pc]; self.pc += 1
                self.stack.append(self.constants[idx])
            elif op == HALT:
                return self.stack[-1] if self.stack else None
            elif op == PUSH:
                self.stack.append(code[self.pc]); self.pc += 1
            elif op == ADD:
                b = self.stack.pop(); self.stack[-1] += b
            elif op == SUB:
                b = self.stack.pop(); self.stack[-1] -= b
            elif op == MUL:
                b = self.stack.pop(); self.stack[-1] *= b
            elif op == DIV:
                b = self.stack.pop(); self.stack[-1] //= b
            elif op == EQ:
                b = self.stack.pop(); self.stack[-1] = int(self.stack[-1] == b)
            elif op == LT:
                b = self.stack.pop(); self.stack[-1] = int(self.stack[-1] < b)
            elif op == NOT:
                self.stack[-1] = int(not self.stack[-1])
            elif op == JMP:
                self.pc = self._u16()
            elif op == JMPZ:
                target = self._u16()
                if not self.stack.pop(): self.pc = target
            elif op == LOAD:
                s = code[self.pc]; self.pc += 1; self.stack.append(self.locals[s])
            elif op == STORE:
                s = code[self.pc]; self.pc += 1; self.locals[s] = self.stack.pop()
            elif op == CALL:
                target = self._u16()
                self.frames.append((self.locals[:], self.pc))
                self.locals = [0] * 16
                self.pc = target
            elif op == RET:
                val = self.stack.pop()
                self.locals, self.pc = self.frames.pop()
                self.stack.append(val)
            else:
                raise ValueError(f"unknown opcode {op} at pc={self.pc - 1}")


# -- Demo

if __name__ == "__main__":
    # 3 * (4 + 5) = 27
    prog = [
        PUSH, 3,
        PUSH, 4,
        PUSH, 5,
        ADD,
        MUL,
        HALT,
    ]
    result = StackVM(prog).run()
    assert result == 27
    print(f"3 * (4 + 5) = {result}")

    # factorial(5) = 120
    #
    # Address layout:
    #   0  PUSH 5       bootstrap: push argument
    #   2  CALL [6]     call fact
    #   5  HALT
    #   6  STORE 0      fact: n -> locals[0]
    #   8  LOAD  0      push n
    #  10  PUSH  1
    #  12  LT           n < 1?
    #  13  NOT          flip: 1 if n >= 1
    #  14  JMPZ [29]    if n == 0, go to base case
    #  17  LOAD  0      push n  (stays on stack across recursive call)
    #  19  LOAD  0      push n again
    #  21  PUSH  1
    #  23  SUB          n - 1
    #  24  CALL  [6]    fact(n-1); result replaces n-1 on stack
    #  27  MUL          n * fact(n-1)
    #  28  RET
    #  29  PUSH  1      base: return 1
    #  31  RET
    fact = [
        PUSH, 5, CALL, 0, 6, HALT,
        STORE, 0, LOAD, 0, PUSH, 1, LT, NOT, JMPZ, 0, 29,
        LOAD, 0, LOAD, 0, PUSH, 1, SUB, CALL, 0, 6, MUL, RET,
        PUSH, 1, RET,
    ]
    result = StackVM(fact).run()
    assert result == 120
    print(f"factorial(5) = {result}")

    # Constant pool: load a value wider than one byte
    constants = [1000]
    pool_prog = [
        LOADK, 0,   # push constants[0] = 1000
        LOADK, 0,   # push 1000 again
        ADD,        # 2000
        HALT,
    ]
    result = StackVMWithPool(pool_prog, constants).run()
    assert result == 2000
    print(f"1000 + 1000 (via constant pool) = {result}")

File: ch02/c/stack_vm.c
/*
 * stack_vm.c — a C parallel of ch02/stack_vm.py.
 *
 * The same fifteen-instruction stack machine, implemented THREE ways so the
 * dispatch discussion of Section 2.4 (sec:vm-dispatch) can be measured rather
 * than asserted:
 *
 *   run_cascade  — an if/else-if comparison chain   (the C analogue of Python's
 *                  if/elif; O(n) comparisons per opcode)
 *   run_switch   — a dense switch, which the compiler turns into a jump table
 *   run_goto     — computed goto (a GCC/Clang extension): each handler ends with
 *                  its own indirect branch, giving the branch predictor more to
 *                  work with
 *
 * All three share one set of operation macros, so their semantics are identical
 * by construction — including Python's floor division for DIV, so this VM is a
 * faithful differential twin of stack_vm.py: the same bytecode yields the same
 * result in both. See solutions.md (Exercise 5) for the cross-check.
 *
 * Build & run:   make run            (samples + benchmark at N = 20,000,000)
 *                ./stack_vm 1000000  (choose N)
 *                make check          (asserts correctness, all dispatchers agree)
 */

#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

/* Opcodes — identical numbering to stack_vm.py (HALT = 0 … RET = 14). */
enum {
    HALT, PUSH, ADD, SUB, MUL, DIV, EQ, LT, NOT,
    JMP, JMPZ, LOAD, STORE, CALL, RET
};

#define STACK_MAX 65536
#define FRAME_MAX 4096

typedef struct { int64_t locals[16]; int ret_pc; } Frame;

/* Python's // rounds toward negative infinity; C's / truncates toward zero.
 * Match Python so the two VMs agree on mixed-sign division (the chapter's
 * "one simplification to flag"). */
static inline int64_t floordiv(int64_t a, int64_t b) {
    int64_t q = a / b, r = a % b;
    if (r != 0 && ((r < 0) != (b < 0))) q--;
    return q;
}

/* One definition of each opcode's effect, shared by all three dispatchers.
 * They read/write the locals declared in each run_* function: code, pc, sp,
 * stack, locals, frames, fp. */
#define RD_U16()  ((uint16_t)((code[pc] << 8) | code[pc + 1]))

#define DO_PUSH   (stack[++sp] = code[pc++])
#define DO_ADD    do { int64_t b = stack[sp--]; stack[sp] += b; } while (0)
#define DO_SUB    do { int64_t b = stack[sp--]; stack[sp] -= b; } while (0)
#define DO_MUL    do { int64_t b = stack[sp--]; stack[sp] *= b; } while (0)
#define DO_DIV    do { int64_t b = stack[sp--]; stack[sp] = floordiv(stack[sp], b); } while (0)
#define DO_EQ     do { int64_t b = stack[sp--]; stack[sp] = (stack[sp] == b); } while (0)
#define DO_LT     do { int64_t b = stack[sp--]; stack[sp] = (stack[sp] <  b); } while (0)
#define DO_NOT    (stack[sp] = !stack[sp])
#define DO_JMP    do { pc = RD_U16(); } while (0)
#define DO_JMPZ   do { uint16_t t = RD_U16(); pc += 2; if (stack[sp--] == 0) pc = t; } while (0)
#define DO_LOAD   do { uint8_t s = code[pc++]; stack[++sp] = locals[s]; } while (0)
#define DO_STORE  do { uint8_t s = code[pc++]; locals[s] = stack[sp--]; } while (0)
#define DO_CALL   do { uint16_t t = RD_U16(); pc += 2;                     \
                       memcpy(frames[fp].locals, locals, sizeof locals);   \
                       frames[fp].ret_pc = pc; fp++;                       \
                       memset(locals, 0, sizeof locals); pc = t; } while (0)
#define DO_RET    do { int64_t v = stack[sp--]; fp--;                      \
                       memcpy(locals, frames[fp].locals, sizeof locals);   \
                       pc = frames[fp].ret_pc; stack[++sp] = v; } while (0)

#define DECLARE_STATE                                  \
    int64_t stack[STACK_MAX]; int sp = -1;             \
    int64_t locals[16] = {0}; locals[0] = init0;       \
    Frame frames[FRAME_MAX]; int fp = 0;               \
    int pc = 0

/* ── 1. comparison cascade ──────────────────────────────────────────────── */
static int64_t run_cascade(const uint8_t *code, int64_t init0) {
    DECLARE_STATE;
    for (;;) {
        uint8_t op = code[pc++];
        if      (op == HALT)  return sp >= 0 ? stack[sp] : 0;
        else if (op == PUSH)  DO_PUSH;
        else if (op == ADD)   DO_ADD;
        else if (op == SUB)   DO_SUB;
        else if (op == MUL)   DO_MUL;
        else if (op == DIV)   DO_DIV;
        else if (op == EQ)    DO_EQ;
        else if (op == LT)    DO_LT;
        else if (op == NOT)   DO_NOT;
        else if (op == JMP)   DO_JMP;
        else if (op == JMPZ)  DO_JMPZ;
        else if (op == LOAD)  DO_LOAD;
        else if (op == STORE) DO_STORE;
        else if (op == CALL)  DO_CALL;
        else if (op == RET)   DO_RET;
        else { fprintf(stderr, "bad op %d at pc=%d\n", op, pc - 1); exit(1); }
    }
}

/* ── 2. switch (compiler builds a jump table) ───────────────────────────── */
static int64_t run_switch(const uint8_t *code, int64_t init0) {
    DECLARE_STATE;
    for (;;) {
        uint8_t op = code[pc++];
        switch (op) {
            case HALT:  return sp >= 0 ? stack[sp] : 0;
            case PUSH:  DO_PUSH;  break;
            case ADD:   DO_ADD;   break;
            case SUB:   DO_SUB;   break;
            case MUL:   DO_MUL;   break;
            case DIV:   DO_DIV;   break;
            case EQ:    DO_EQ;    break;
            case LT:    DO_LT;    break;
            case NOT:   DO_NOT;   break;
            case JMP:   DO_JMP;   break;
            case JMPZ:  DO_JMPZ;  break;
            case LOAD:  DO_LOAD;  break;
            case STORE: DO_STORE; break;
            case CALL:  DO_CALL;  break;
            case RET:   DO_RET;   break;
            default: fprintf(stderr, "bad op %d at pc=%d\n", op, pc - 1); exit(1);
        }
    }
}

/* ── 3. computed goto (GCC/Clang) ───────────────────────────────────────── */
#if defined(__GNUC__)
#define HAVE_COMPUTED_GOTO 1
static int64_t run_goto(const uint8_t *code, int64_t init0) {
    DECLARE_STATE;
    void *table[] = {
        &&do_halt, &&do_push, &&do_add, &&do_sub, &&do_mul, &&do_div,
        &&do_eq, &&do_lt, &&do_not, &&do_jmp, &&do_jmpz, &&do_load,
        &&do_store, &&do_call, &&do_ret
    };
    #define NEXT() goto *table[code[pc++]]
    NEXT();
    do_halt:  return sp >= 0 ? stack[sp] : 0;
    do_push:  DO_PUSH;  NEXT();
    do_add:   DO_ADD;   NEXT();
    do_sub:   DO_SUB;   NEXT();
    do_mul:   DO_MUL;   NEXT();
    do_div:   DO_DIV;   NEXT();
    do_eq:    DO_EQ;    NEXT();
    do_lt:    DO_LT;    NEXT();
    do_not:   DO_NOT;   NEXT();
    do_jmp:   DO_JMP;   NEXT();
    do_jmpz:  DO_JMPZ;  NEXT();
    do_load:  DO_LOAD;  NEXT();
    do_store: DO_STORE; NEXT();
    do_call:  DO_CALL;  NEXT();
    do_ret:   DO_RET;   NEXT();
    #undef NEXT
}
#else
#define HAVE_COMPUTED_GOTO 0
static int64_t run_goto(const uint8_t *code, int64_t init0) {
    return run_switch(code, init0);   /* fallback where the extension is absent */
}
#endif

/* ── Sample programs (same bytecode the Python VM runs) ──────────────────── */

/* 3 * (4 + 5) = 27 */
static const uint8_t prog_mul[] = {
    PUSH, 3, PUSH, 4, PUSH, 5, ADD, MUL, HALT
};

/* factorial(5) = 120 (identical layout to stack_vm.py's `fact`) */
static const uint8_t prog_fact[] = {
    PUSH, 5, CALL, 0, 6, HALT,
    STORE, 0, LOAD, 0, PUSH, 1, LT, NOT, JMPZ, 0, 29,
    LOAD, 0, LOAD, 0, PUSH, 1, SUB, CALL, 0, 6, MUL, RET,
    PUSH, 1, RET
};

/* sum 1..N : locals[0] = N (preset via init0), locals[1] = acc.
 * Same bytecode as solutions/ex05_dispatch.py (end @ 25). */
static const uint8_t prog_sum[] = {
    /* 0  loop */ PUSH, 0, LOAD, 0, LT,
    /* 5  */      JMPZ, 0, 25,
    /* 8  */      LOAD, 1, LOAD, 0, ADD, STORE, 1,
    /* 15 */      LOAD, 0, PUSH, 1, SUB, STORE, 0,
    /* 22 */      JMP, 0, 0,
    /* 25 end */  LOAD, 1, HALT
};

/* ── Benchmark helper ───────────────────────────────────────────────────── */
typedef int64_t (*run_fn)(const uint8_t *, int64_t);

static double bench(run_fn fn, const uint8_t *code, int64_t N, int reps) {
    double best = 1e30;
    for (int i = 0; i < reps; i++) {
        struct timespec a, b;
        clock_gettime(CLOCK_MONOTONIC, &a);
        volatile int64_t r = fn(code, N); (void) r;
        clock_gettime(CLOCK_MONOTONIC, &b);
        double t = (b.tv_sec - a.tv_sec) + (b.tv_nsec - a.tv_nsec) / 1e9;
        if (t < best) best = t;
    }
    return best;
}

int main(int argc, char **argv) {
    long N = argc > 1 ? atol(argv[1]) : 20000000L;

    /* correctness — these must match stack_vm.py exactly */
    int64_t r_mul  = run_switch(prog_mul, 0);
    int64_t r_fact = run_switch(prog_fact, 0);
    if (r_mul != 27 || r_fact != 120) {
        fprintf(stderr, "FAIL samples: %lld %lld\n",
                (long long) r_mul, (long long) r_fact);
        return 1;
    }
    printf("3 * (4 + 5)  = %lld\n", (long long) r_mul);
    printf("factorial(5) = %lld\n", (long long) r_fact);

    /* all three dispatchers must agree on the sum (differential within C) */
    int64_t expected = (int64_t) N * (N + 1) / 2;
    int64_t s_casc = run_cascade(prog_sum, N);
    int64_t s_sw   = run_switch(prog_sum, N);
    int64_t s_go   = run_goto(prog_sum, N);
    if (s_casc != expected || s_sw != expected || s_go != expected) {
        fprintf(stderr, "FAIL sum: cascade=%lld switch=%lld goto=%lld expected=%lld\n",
                (long long) s_casc, (long long) s_sw,
                (long long) s_go, (long long) expected);
        return 1;
    }
    printf("sum 1..%ld = %lld  (cascade == switch == goto)\n",
           N, (long long) expected);

    /* benchmark: best of 3 */
    double t_casc = bench(run_cascade, prog_sum, N, 3);
    double t_sw   = bench(run_switch,  prog_sum, N, 3);
    double t_go   = bench(run_goto,    prog_sum, N, 3);

    printf("\ndispatch on sum 1..%ld (best of 3, lower is better):\n", N);
    printf("  %-20s%10s\n", "strategy", "seconds");
    printf("  %-20s%10.3f\n", "comparison cascade", t_casc);
    printf("  %-20s%10.3f\n", "switch", t_sw);
    printf("  %-20s%10.3f%s\n", "computed goto", t_go,
           HAVE_COMPUTED_GOTO ? "" : "  (fell back to switch)");
    return 0;
}
